stan/math/rev/meta/rev_matrix_type.hpp

```diff
@@ -0,0 +1,38 @@
+#ifndef STAN_MATH_REV_META_REV_MATRIX_TYPE_HPP
+#define STAN_MATH_REV_META_REV_MATRIX_TYPE_HPP
+
+#include <stan/math/prim/meta/disjunction.hpp>
+#include <stan/math/prim/meta/conjunction.hpp>
+#include <stan/math/prim/meta/require_generics.hpp>
+#include <stan/math/prim/meta/return_type.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/rev/core/var.hpp>
+
+namespace stan {
+
+/**
+ * Determines a return type for a function that accepts given inputs and wants
+ * to return a matrix (or vector or row vector) with given compile time number
+ * of rows and columns.
+ *
+ * Returns `var_value` if any of the inputs are `var_value`, otherwise an Eigen
+ * matrix of `var` or `double`.
+ * @tparam Rows number of compile time rows
+ * @tparam Cols number of compile time columns
+ * @tparam Inputs types of arguments the function accepts
+ */
+template <int Rows, int Cols, typename... Inputs>
+struct rev_matrix_type {
+  using type = std::conditional_t<
+      math::disjunction<math::conjunction<
+          is_var<Inputs>, is_eigen<value_type_t<Inputs>>>...>::value,
+      math::var_value<Eigen::Matrix<double, Rows, Cols>>,
+      Eigen::Matrix<return_type_t<Inputs...>, Rows, Cols>>;
+};
```

Comment:
So it looks like it's decision time on whether functions that can operator on mixes of `var<mat>` and `mat<var>` should return a `var<mat>` or `mat<var>`. 

I agree with the more aggressive version here where `var<mat>` is returned by default. For most problems it feels like having the reverse mode pass using `var<mat>` would help offset the copy cost when we need to go back and forth between the two for some functions. @bbbales2 @rok-cesnovar do yinz have opinions on this as well?

Reply:
I am very for returning `var_value<Eigen>` as much as possible.

If `matrix[N, M]` in the language stays corresponding to an `Eigen::Matrix<var, R, C>` with no compiler fanciness, then we'll induce in reverse mode an expensive copy every time we copy a `var_value` to one of these.

What makes this expensive is that to do propagate the adjoints back we need to:
1. Copy out the adjoints of the `Eigen::Matrix<var, R, C>` into the adjoint of the `var_value<Matrix>`
2. Read the adjoint of the `var_value<Matrix>` and propagate stuff back from there

This is compared to the cheap operation:
1. Read the adjoints of `Eigen::Matrix<var, R, C>` and propagate stuff back from there.

Saw this amusing thing the other day -- I think it applies: [cigarettes and cocaine](http://observationalepidemiology.blogspot.com/2020/08/a-cigarettes-and-cocaine-argument.html). We're cutting the cigarettes with this optimization -- the pointer chasing in reading the adjoints of `Eigen::Matrix<var, R, C>` is the expensive thing and we're doing it either way.

And the upside here is that returning `var_value<Matrix>` variables, we'll have more of them floating around and can get efficiencies in basically any situation other than assigning to `Eigen::Matrix<var, R, C>` types.

Reply:
Word then I think this code is A-OK!